File: ios/Classes/Viewport/ViewportCalculator.swift
import CoreGraphics
import Foundation

struct ResolvedCrosshair {
  let candleIndex: Int
  let snappedX: CGFloat
  let y: CGFloat
}

final class ViewportCalculator {
  private(set) var visibleStartIndex: Int = 0
  private(set) var visibleEndIndex: Int = -1
  private(set) var visibleCandleCount: Int = 0
  private(set) var candleWidth: Double = 0
  private(set) var slotWidth: Double = 0
  private(set) var priceHigh: Double = 0
  private(set) var priceLow: Double = 0
  private(set) var maxVisibleVolume: Double = 0
  private(set) var isAtLatest: Bool = true
  private var viewportStart: Double = 0
  private var visibleCountDouble: Double = 0

  /// Fractional scroll position in candle index space (for time labels that track pan smoothly).
  var scrollOffset: Double { viewportStart }
  private var scaleBaseStart: Double?
  private var scaleBaseVisibleCount: Double?

  /// Below this slot width (logical points), candle mode renders as a close line (Bybit-style LOD).
  private static let candleLodMaxSlotWidth: Double = 4.0

  func resetToLatest(
    candleStore: CandleStore,
    plotWidth: Double,
    config: ConfigMessage,
    chartType: String
  ) {
    guard candleStore.count > 0, plotWidth > 0 else {
      clear()
      return
    }

    visibleCountDouble = Double(calculateVisibleCount(
      candleCount: candleStore.count,
      plotWidth: plotWidth,
      config: config
    ))
    viewportStart = maxStart(candleCount: candleStore.count)
    updateDerived(
      candleStore: candleStore,
      plotWidth: plotWidth,
      config: config,
      chartType: chartType
    )
  }

  func refresh(
    candleStore: CandleStore,
    plotWidth: Double,
    config: ConfigMessage,
    chartType: String
  ) {
    guard candleStore.count > 0, plotWidth > 0 else {
      clear()
      return
    }

    if visibleCountDouble <= 0 {
      resetToLatest(
        candleStore: candleStore,
        plotWidth: plotWidth,
        config: config,
        chartType: chartType
      )
      return
    }

    visibleCountDouble = min(
      Double(candleStore.count),
      min(Double(config.maxVisibleCandles), max(Double(config.minVisibleCandles), visibleCountDouble))
    )
    viewportStart = min(max(0, viewportStart), maxStart(candleCount: candleStore.count))
    updateDerived(
      candleStore: candleStore,
      plotWidth: plotWidth,
      config: config,
      chartType: chartType
    )
  }

  func hasVisibleRange() -> Bool {
    visibleCandleCount > 0 && visibleEndIndex >= visibleStartIndex
  }

  /// Renders candle mode as a line when zoomed out (matches price-range logic in `updateDerived`).
  func effectiveRenderChartType(userChartType: String) -> String {
    Self.rangeChartTypeForPriceBounds(chartType: userChartType, slotWidth: slotWidth)
  }

  /// Candle index for a fraction of the visible span (0 = left edge, 1 = right), for time-axis labels.
  func candleIndexAtVisibleRatio(_ t: Double, candleCount: Int) -> Int {
    guard hasVisibleRange(), candleCount > 0, visibleCountDouble > 0 else {
      return 0
    }
    let u = min(1.0, max(0.0, t))
    let raw = viewportStart + u * visibleCountDouble - 0.5
    let idx = Int(round(raw))
    return min(max(0, idx), candleCount - 1)
  }

  func xCenter(for index: Int, plotLeft: CGFloat) -> CGFloat {
    plotLeft + CGFloat((Double(index) - viewportStart) + 0.5) * CGFloat(slotWidth)
  }

  func candleBodyWidth() -> CGFloat {
    CGFloat(candleWidth)
  }

  func priceToY(_ price: Double, top: CGFloat, height: CGFloat) -> CGFloat {
    let range = priceHigh - priceLow > 0 ? priceHigh - priceLow : 1
    let normalized = CGFloat((priceHigh - price) / range)
    return top + normalized * height
  }

  func volumeToY(_ volume: Double, top: CGFloat, height: CGFloat) -> CGFloat {
    let normalized: CGFloat
    if maxVisibleVolume <= 0 {
      normalized = 0
    } else {
      normalized = CGFloat(volume / maxVisibleVolume)
    }
    return top + height - normalized * height
  }

  func applyPanDelta(
    deltaX: Double,
    candleStore: CandleStore,
    plotWidth: Double,
    config: ConfigMessage,
    chartType: String
  ) -> Bool {
    guard hasVisibleRange(), slotWidth > 0 else {
      return false
    }

    let newStart = min(
      max(0, viewportStart - (deltaX / slotWidth)),
      maxStart(candleCount: candleStore.count)
    )
    guard abs(newStart - viewportStart) > 0.0001 else {
      return false
    }

    viewportStart = newStart
    updateDerived(
      candleStore: candleStore,
      plotWidth: plotWidth,
      config: config,
      chartType: chartType
    )
    return true
  }

  func applyScale(
    scaleFactor: Double,
    focalPointX: Double,
    candleStore: CandleStore,
    plotWidth: Double,
    config: ConfigMessage,
    chartType: String
  ) -> Bool {
    guard candleStore.count > 0, plotWidth > 0 else {
      return false
    }

    if scaleBaseVisibleCount == nil {
      scaleBaseVisibleCount = visibleCountDouble
      scaleBaseStart = viewportStart
    }

    let baseVisibleCount = scaleBaseVisibleCount ?? visibleCountDouble
    let baseStart = scaleBaseStart ?? viewportStart
    let clampedScale = max(scaleFactor, 0.2)
    let newVisibleCount = min(
      Double(candleStore.count),
      min(
        Double(config.maxVisibleCandles),
        max(Double(config.minVisibleCandles), baseVisibleCount / clampedScale)
      )
    )
    let focalRatio = min(max(0, focalPointX / plotWidth), 1)
    let focalCandle = baseStart + (focalRatio * baseVisibleCount)
    let newStart = min(
      max(0, focalCandle - (focalRatio * newVisibleCount)),
      max(0, Double(candleStore.count) - newVisibleCount)
    )

    let changed = abs(newVisibleCount - visibleCountDouble) > 0.0001 || abs(newStart - viewportStart) > 0.0001
    visibleCountDouble = newVisibleCount
    viewportStart = newStart
    updateDerived(
      candleStore: candleStore,
      plotWidth: plotWidth,
      config: config,
      chartType: chartType
    )
    return changed
  }

  func endScaleGesture() {
    scaleBaseStart = nil
    scaleBaseVisibleCount = nil
  }

  func scrollToEnd(
    candleStore: CandleStore,
    plotWidth: Double,
    config: ConfigMessage,
    chartType: String
  ) -> Bool {
    guard candleStore.count > 0 else {
      return false
    }

    let newStart = maxStart(candleCount: candleStore.count)
    guard abs(newStart - viewportStart) > 0.0001 else {
      return false
    }

    viewportStart = newStart
    updateDerived(
      candleStore: candleStore,
      plotWidth: plotWidth,
      config: config,
      chartType: chartType
    )
    return true
  }

  func resolveCrosshair(
    x: Double,
    y: Double,
    contentRect: CGRect,
    mainChartRect: CGRect,
    candleStore: CandleStore
  ) -> ResolvedCrosshair? {
    guard hasVisibleRange(), slotWidth > 0 else {
      return nil
    }

    let clampedX = min(max(contentRect.minX, CGFloat(x)), contentRect.maxX)
    let clampedY = min(max(mainChartRect.minY, CGFloat(y)), mainChartRect.maxY)
    let projectedIndex = Int(
      round(viewportStart + (Double(clampedX - contentRect.minX) / slotWidth) - 0.5)
    )
    let index = min(max(visibleStartIndex, projectedIndex), visibleEndIndex)

    return ResolvedCrosshair(
      candleIndex: min(max(0, index), candleStore.count - 1),
      snappedX: xCenter(for: index, plotLeft: contentRect.minX),
      y: clampedY
    )
  }

  private func updateDerived(
    candleStore: CandleStore,
    plotWidth: Double,
    config: ConfigMessage,
    chartType: String
  ) {
    visibleStartIndex = max(0, Int(floor(viewportStart)))
    visibleEndIndex = min(candleStore.count - 1, Int(ceil(viewportStart + visibleCountDouble)) - 1)
    visibleCandleCount = max(0, visibleEndIndex - visibleStartIndex + 1)
    slotWidth = plotWidth / visibleCountDouble
    candleWidth = max(1, slotWidth * 0.72)
    isAtLatest = viewportStart >= maxStart(candleCount: candleStore.count) - 0.01
    let rangeChartType = Self.rangeChartTypeForPriceBounds(
      chartType: chartType,
      slotWidth: slotWidth
    )
    recalculateRanges(
      candleStore: candleStore,
      config: config,
      chartType: rangeChartType
    )
  }

  private static func rangeChartTypeForPriceBounds(chartType: String, slotWidth: Double) -> String {
    if chartType == "candle", slotWidth > 0, slotWidth < candleLodMaxSlotWidth {
      return "line"
    }
    return chartType
  }

  private func recalculateRanges(
    candleStore: CandleStore,
    config: ConfigMessage,
    chartType: String
  ) {
    guard hasVisibleRange() else {
      priceHigh = 0
      priceLow = 0
      maxVisibleVolume = 0
      return
    }

    var minPrice = Double.greatestFiniteMagnitude
    var maxPrice = -Double.greatestFiniteMagnitude
    var maxVolume = 0.0

    for index in visibleStartIndex...visibleEndIndex {
      let high = chartType == "line" ? candleStore.close(at: index) : candleStore.high(at: index)
      let low = chartType == "line" ? candleStore.close(at: index) : candleStore.low(at: index)
      minPrice = Swift.min(minPrice, low)
      maxPrice = Swift.max(maxPrice, high)
      maxVolume = Swift.max(maxVolume, candleStore.volume(at: index))
    }

    if minPrice == Double.greatestFiniteMagnitude || maxPrice == -Double.greatestFiniteMagnitude {
      minPrice = 0
      maxPrice = 1
    }

    if minPrice == maxPrice {
      let padding = minPrice == 0 ? 1 : minPrice * config.yAxisPaddingRatio
      minPrice -= padding
      maxPrice += padding
    } else {
      let padding = (maxPrice - minPrice) * config.yAxisPaddingRatio
      minPrice -= padding
      maxPrice += padding
    }

    priceLow = minPrice
    priceHigh = maxPrice
    maxVisibleVolume = maxVolume
  }

  private func calculateVisibleCount(
    candleCount: Int,
    plotWidth: Double,
    config: ConfigMessage
  ) -> Int {
    let idealSlotWidth = 10.0
    let estimated = max(1, Int(floor(plotWidth / idealSlotWidth)))
    let clamped = min(
      Int(config.maxVisibleCandles),
      max(Int(config.minVisibleCandles), estimated)
    )
    return min(candleCount, clamped)
  }

  private func maxStart(candleCount: Int) -> Double {
    max(0, Double(candleCount) - visibleCountDouble)
  }

  private func clear() {
    visibleStartIndex = 0
    visibleEndIndex = -1
    visibleCandleCount = 0
    candleWidth = 0
    slotWidth = 0
    priceHigh = 0
    priceLow = 0
    maxVisibleVolume = 0
    isAtLatest = true
    viewportStart = 0
    visibleCountDouble = 0
    endScaleGesture()
  }
}

File: android/src/main/kotlin/com/tradechart/plugin/viewport/ViewportCalculator.kt
package com.tradechart.plugin.viewport

import android.graphics.RectF
import com.tradechart.plugin.bridge.generated.ConfigMessage
import com.tradechart.plugin.data.CandleStore
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/** Below this slot width (logical px), candle mode draws as a close line (Bybit-style LOD). */
private const val CANDLE_LOD_MAX_SLOT_WIDTH = 4.0

data class ResolvedCrosshair(
    val candleIndex: Int,
    val snappedX: Float,
    val y: Float,
)

class ViewportCalculator {
    var visibleStartIndex: Int = 0
        private set
    var visibleEndIndex: Int = -1
        private set
    var visibleCandleCount: Int = 0
        private set
    var candleWidth: Double = 0.0
        private set
    var slotWidth: Double = 0.0
        private set
    var priceHigh: Double = 0.0
        private set
    var priceLow: Double = 0.0
        private set
    var maxVisibleVolume: Double = 0.0
        private set
    var isAtLatest: Boolean = true
        private set

    private var viewportStart: Double = 0.0
    private var visibleCountDouble: Double = 0.0

    val scrollOffset: Double
        get() = viewportStart
    private var scaleBaseStart: Double? = null
    private var scaleBaseVisibleCount: Double? = null

    private fun rangeChartTypeForPriceBounds(chartType: String, slotWidth: Double): String {
        if (chartType == "candle" && slotWidth > 0.0 && slotWidth < CANDLE_LOD_MAX_SLOT_WIDTH) {
            return "line"
        }
        return chartType
    }

    fun resetToLatest(
        candleStore: CandleStore,
        plotWidth: Double,
        config: ConfigMessage,
        chartType: String,
    ) {
        if (candleStore.count == 0 || plotWidth <= 0.0) {
            clear()
            return
        }
        visibleCountDouble = calculateVisibleCount(
            candleCount = candleStore.count,
            plotWidth = plotWidth,
            config = config,
        ).toDouble()
        viewportStart = maxStart(candleStore.count)
        updateDerived(candleStore, plotWidth, config, chartType)
    }

    fun refresh(
        candleStore: CandleStore,
        plotWidth: Double,
        config: ConfigMessage,
        chartType: String,
    ) {
        if (candleStore.count == 0 || plotWidth <= 0.0) {
            clear()
            return
        }
        if (visibleCountDouble <= 0.0) {
            resetToLatest(candleStore, plotWidth, config, chartType)
            return
        }
        visibleCountDouble = visibleCountDouble
            .coerceAtLeast(config.minVisibleCandles.toDouble())
            .coerceAtMost(config.maxVisibleCandles.toDouble())
            .coerceAtMost(candleStore.count.toDouble())
        viewportStart = viewportStart.coerceIn(0.0, maxStart(candleStore.count))
        updateDerived(candleStore, plotWidth, config, chartType)
    }

    fun hasVisibleRange(): Boolean {
        return visibleCandleCount > 0 && visibleEndIndex >= visibleStartIndex
    }

    fun effectiveRenderChartType(userChartType: String): String {
        return rangeChartTypeForPriceBounds(userChartType, slotWidth)
    }

    fun candleIndexAtVisibleRatio(t: Double, candleCount: Int): Int {
        if (!hasVisibleRange() || candleCount <= 0 || visibleCountDouble <= 0.0) {
            return 0
        }
        val u = t.coerceIn(0.0, 1.0)
        val raw = viewportStart + u * visibleCountDouble - 0.5
        val idx = raw.roundToInt()
        return idx.coerceIn(0, candleCount - 1)
    }

    fun xCenterForIndex(index: Int, plotLeft: Float): Float {
        return plotLeft + ((index - viewportStart) + 0.5) .toFloat() * slotWidth.toFloat()
    }

    fun candleBodyWidth(): Float = candleWidth.toFloat()

    fun priceToY(price: Double, top: Float, height: Float): Float {
        val range = (priceHigh - priceLow).takeIf { it > 0.0 } ?: 1.0
        val normalized = ((priceHigh - price) / range).toFloat()
        return top + normalized * height
    }

    fun volumeToY(volume: Double, top: Float, height: Float): Float {
        val normalized = if (maxVisibleVolume <= 0.0) {
            0.0f
        } else {
            (volume / maxVisibleVolume).toFloat()
        }
        return top + height - normalized * height
    }

    fun applyPanDelta(
        deltaX: Double,
        candleStore: CandleStore,
        plotWidth: Double,
        config: ConfigMessage,
        chartType: String,
    ): Boolean {
        if (!hasVisibleRange() || slotWidth <= 0.0) {
            return false
        }
        val newStart = (viewportStart - (deltaX / slotWidth)).coerceIn(0.0, maxStart(candleStore.count))
        if (newStart == viewportStart) {
            return false
        }
        viewportStart = newStart
        updateDerived(candleStore, plotWidth, config, chartType)
        return true
    }

    fun applyScale(
        scaleFactor: Double,
        focalPointX: Double,
        candleStore: CandleStore,
        plotWidth: Double,
        config: ConfigMessage,
        chartType: String,
    ): Boolean {
        if (candleStore.count == 0 || plotWidth <= 0.0) {
            return false
        }
        val baseVisibleCount = scaleBaseVisibleCount ?: visibleCountDouble.also {
            scaleBaseVisibleCount = it
            scaleBaseStart = viewportStart
        }
        val baseStart = scaleBaseStart ?: viewportStart
        val clampedScale = scaleFactor.coerceAtLeast(0.2)
        val newVisibleCount = (baseVisibleCount / clampedScale)
            .coerceAtLeast(config.minVisibleCandles.toDouble())
            .coerceAtMost(config.maxVisibleCandles.toDouble())
            .coerceAtMost(candleStore.count.toDouble())
        val focalRatio = (focalPointX / plotWidth).coerceIn(0.0, 1.0)
        val focalCandle = baseStart + focalRatio * baseVisibleCount
        val newStart = (focalCandle - focalRatio * newVisibleCount).coerceIn(
            0.0,
            max(0.0, candleStore.count - newVisibleCount),
        )
        val changed = newVisibleCount != visibleCountDouble || newStart != viewportStart
        visibleCountDouble = newVisibleCount
        viewportStart = newStart
        updateDerived(candleStore, plotWidth, config, chartType)
        return changed
    }

    fun endScaleGesture() {
        scaleBaseStart = null
        scaleBaseVisibleCount = null
    }

    fun scrollToEnd(
        candleStore: CandleStore,
        plotWidth: Double,
        config: ConfigMessage,
        chartType: String,
    ): Boolean {
        if (candleStore.count == 0) {
            return false
        }
        val newStart = maxStart(candleStore.count)
        if (newStart == viewportStart) {
            return false
        }
        viewportStart = newStart
        updateDerived(candleStore, plotWidth, config, chartType)
        return true
    }

    fun resolveCrosshair(
        x: Double,
        y: Double,
        contentRect: RectF,
        mainChartRect: RectF,
        candleStore: CandleStore,
    ): ResolvedCrosshair? {
        if (!hasVisibleRange() || slotWidth <= 0.0) {
            return null
        }
        val clampedX = x.coerceIn(contentRect.left.toDouble(), contentRect.right.toDouble())
        val clampedY = y.coerceIn(mainChartRect.top.toDouble(), mainChartRect.bottom.toDouble())
        val index = (viewportStart + ((clampedX - contentRect.left) / slotWidth) - 0.5)
            .roundToInt()
            .coerceIn(visibleStartIndex, visibleEndIndex)
        return ResolvedCrosshair(
            candleIndex = index.coerceIn(0, candleStore.count - 1),
            snappedX = xCenterForIndex(index, contentRect.left),
            y = clampedY.toFloat(),
        )
    }

    private fun updateDerived(
        candleStore: CandleStore,
        plotWidth: Double,
        config: ConfigMessage,
        chartType: String,
    ) {
        visibleStartIndex = floor(viewportStart).toInt().coerceAtLeast(0)
        visibleEndIndex = min(
            candleStore.count - 1,
            ceil(viewportStart + visibleCountDouble).toInt() - 1,
        )
        visibleCandleCount = max(0, visibleEndIndex - visibleStartIndex + 1)
        slotWidth = plotWidth / visibleCountDouble
        candleWidth = max(1.0, slotWidth * 0.72)
        isAtLatest = viewportStart >= maxStart(candleStore.count) - 0.01
        val rangeChartType = rangeChartTypeForPriceBounds(chartType, slotWidth)
        recalculateRanges(candleStore, config, rangeChartType)
    }

    private fun recalculateRanges(
        candleStore: CandleStore,
        config: ConfigMessage,
        chartType: String,
    ) {
        if (!hasVisibleRange()) {
            priceHigh = 0.0
            priceLow = 0.0
            maxVisibleVolume = 0.0
            return
        }

        var minPrice = Double.POSITIVE_INFINITY
        var maxPrice = Double.NEGATIVE_INFINITY
        var maxVolume = 0.0

        for (index in visibleStartIndex..visibleEndIndex) {
            val high = if (chartType == "line") candleStore.closeAt(index) else candleStore.highAt(index)
            val low = if (chartType == "line") candleStore.closeAt(index) else candleStore.lowAt(index)
            minPrice = min(minPrice, low)
            maxPrice = max(maxPrice, high)
            maxVolume = max(maxVolume, candleStore.volumeAt(index))
        }

        if (minPrice == Double.POSITIVE_INFINITY || maxPrice == Double.NEGATIVE_INFINITY) {
            minPrice = 0.0
            maxPrice = 1.0
        }
        if (minPrice == maxPrice) {
            val fallbackPadding = if (minPrice == 0.0) 1.0 else minPrice * config.yAxisPaddingRatio
            minPrice -= fallbackPadding
            maxPrice += fallbackPadding
        } else {
            val padding = (maxPrice - minPrice) * config.yAxisPaddingRatio
            minPrice -= padding
            maxPrice += padding
        }

        priceLow = minPrice
        priceHigh = maxPrice
        maxVisibleVolume = maxVolume
    }

    private fun calculateVisibleCount(
        candleCount: Int,
        plotWidth: Double,
        config: ConfigMessage,
    ): Int {
        val idealSlotWidth = 10.0
        val estimatedCount = floor(plotWidth / idealSlotWidth).toInt().coerceAtLeast(1)
        val clamped = estimatedCount
            .coerceAtLeast(config.minVisibleCandles.toInt())
            .coerceAtMost(config.maxVisibleCandles.toInt())
        return min(candleCount, clamped)
    }

    private fun maxStart(candleCount: Int): Double {
        return max(0.0, candleCount.toDouble() - visibleCountDouble)
    }

    private fun clear() {
        visibleStartIndex = 0
        visibleEndIndex = -1
        visibleCandleCount = 0
        candleWidth = 0.0
        slotWidth = 0.0
        priceHigh = 0.0
        priceLow = 0.0
        maxVisibleVolume = 0.0
        isAtLatest = true
        viewportStart = 0.0
        visibleCountDouble = 0.0
        endScaleGesture()
    }
}
